### ISSUE TITLE
Users management

### ISSUE THREAD

**Issue by piranna** [NodeOS/NodeOS#80]
Now that we have multiuser support, a question arise... how to create them? and who? At this moment the concept of "user" is just an isolated folder on the users filesystem and later each user has the feeling it's the only one on the machine, but how to create that users folders and by who? Who can be able to create new users on the system, anybody? And to delete them?

**Comment by lite20** [comment]
@piranna I'd like to get started on implementing this. First however I have some questions:

How exactly does the NodeOS logon process work? You mention that NodeOS already supports multiple users. How is this achieved? In mounting different user partitions?

In that case the logon process is on the partition itself and simply locks the partition?

**Comment by lite20** [comment]
I'm going to get started on this. I will dump any issues I hit here.

**Comment by lite20** [comment]
> this mechanism is somewhat inspired in how Plan9 works

I love this idea! It would be cool to have your account hosted on a cloud and simply sync to it whenever you want to use it. And now with Yunitto's remote render, you could have your computer hosted remotely in a data center with a Gb per second wifi connection and simply connect via browser from any device.

NodeOS possibilities are unlimited. (maybe I will run said service)

**Comment by lite20** [comment]
> on the users filesystem

Is this current? I thought that a user was simply a mounted layer 4 (userfs), and that multiple users would consist of mounting a different 4th layer.

If this is how it works, than users should be created and managed from the boot process as I assume it is what mounts the userfs layer. This would create completely isolated users and it wouldn't matter who created users and how many they created as they'd all operate independently and ignorantly of the others.

That is to the best of my understanding... which is limited.

**Comment by lite20** [comment]
P.S. feel free to assign me to this issue via github

**Comment by piranna** [comment]
One step at a time, first let's allow `logon` to create new users since it has root permissions, and later we'll see about using remote $home directories (this can be an independent project).

`logon` will only need to check that the username don't exists and ask to create it. If so, create a new folder with that username on the users partition (mounted at `/home` on the root user partition) and a `etc/logon.json` file, and set them and all its files with an already not assigned UID and GID. The user password is stored as a MD5 hash. I think a good first task would be to document all of this on the `logon` README before adding the functionality to create new users, what do you think? You could be able to inspect the source code to learn how it works and document it, so later it will be easier to you to add that functionality :-)

We'll talk about the remote $home other day, I have some ideas about it already... ;-)

P.S.: please don't edit or remove comments after they are published...

**Comment by Coretool** [comment]
>The user password is stored as a MD5 hash 

Isn't it SHA1 ? 
Do you want to add the option of creating a new user to the login dialog ?

**Comment by piranna** [comment]
> The user password is stored as a MD5 hash
> 
> Isn't it SHA1 ?

Err... now I'm not sure :-P I would need to review it in the code, sorry :-P


> Do you want to add the option of creating a new user to the login dialog ?

Yeah, that's the idea: if the user don't exist, ask him about if he want to create a new account.

**Comment by Coretool** [comment]
It is SHA1, I looked it up 😊

**Comment by piranna** [comment]
Cool, thanks :-)

**Comment by Coretool** [comment]
Np, I like correctness (is this even a word ? )

**Comment by piranna** [comment]
> I like correctness (is this even a word ? )

I think so...

**Comment by lite20** [comment]
> > Do you want to add the option of creating a new user to the login dialog ?

> Yeah, that's the idea: if the user don't exist, ask him about if he want to create a new account.

Haha okay I'll see what I can do after school. Do you own the repository of the NodeOS team?

**Comment by Coretool** [comment]
Who ? Me or @piranna ?

**Comment by piranna** [comment]
@lite20 what repo are you talking about? Anyway, maybe better you do a fork and later a pull-request...

**Comment by lite20** [comment]
Sorry, @piranna haha. But I think I've found it. I just wanted to make sure

**Comment by piranna** [comment]
:+1:

**Comment by Coretool** [comment]
Does anybody know why I can't build logon ? While building I get the warning 
`gyp WARN EACCES user "root" does not have permission to access the dev dir "/root/.node-gyp/5.6.0"
`

**Comment by lite20** [comment]
> Does anybody know why I can't build logon ? While building I get the warning 

Weird. Let me ssh to a dev box and try

**Comment by Coretool** [comment]
Its ultra weird... I tried different versions and it fails always...

**Comment by lite20** [comment]
Yep I hit a billion errors at posix too. POSIX is probably legitimately broken. Let's try installing older versions of it

**Comment by piranna** [comment]
Seems you are using `root` user, definitely a bad idea. Remove the `~/.npm` and `~/.node-gyp` directories are try again.

**Comment by lite20** [comment]
> Seems you are using root user, definitely a bad idea. Remove the ~/.npm and ~/.node-gyp directories are try again.

Gah. Root user is the only user this VPS has :P now I gotta make another one just to compile posix

**Comment by Coretool** [comment]
Hm... It's still failing...

**Comment by Coretool** [comment]
Whatever... Is there an example logon.json file anywhere ?

**Comment by piranna** [comment]
> Whatever... Is there an example logon.json file anywhere ?

https://github.com/NodeOS/NodeOS/blob/master/node_modules/nodeos-usersfs/logon.json

**Comment by Coretool** [comment]
Thanks !

**Comment by piranna** [comment]
You are welcome :-)

**Comment by lite20** [comment]
@piranna @Coretool When a new user is created, a few files need to be created in their home directory. Ought we not to create a "blank user" tar file and just unzip that to new home directories on creation? That way we could easily have a base account built quickly without having to embed anything in code.

**Comment by Coretool** [comment]
Hm... This way you can save some typing :P ! I think you should try it out !

**Comment by piranna** [comment]
> @piranna @Coretool When a new user is created, a few files need to be
created in their home directory. Ought we not to create a "blank user" tar
file and just unzip that to new home directories on creation? That way we
could easily have a base account built quickly without having to embed
anything in code.

By design, only needed file is etc/logon.json, but it's true is not very
friendly to give a newcomer a Node.js REPL (although he don't need
more...). We could ask him if he want to install in its account some basic
packages (like nsh) and if so download and copy npm directly from GitHub
(it's ready to use) and use it to install the basic packages. We should
take in account that we can't use compiled modules yet, though...

**Comment by Coretool** [comment]
@piranna, ... i know it doesn't belong in this thread but have you heard of 'cash' ? Maybe we could use it instead of coreutils.js ...

**Comment by Coretool** [comment]
(Or as nsh alternative)

**Comment by piranna** [comment]
Yeah, I have just read about it today, looks pretty interesting but it's only focused on implement the commands, while coreutils.js try to use streams to make them reusable. Also, cash code sucks... :-( Anyway, it's a project that needs to be considerated to see how it evolves :-)

**Comment by Coretool** [comment]
I think we should add 'finish coreutils' to the tasklist :) anyway

**Comment by piranna** [comment]
Definitely, but this will be a long task, and with a solid, working nsh it's mostly useless, that's why I decided to rewrite it.

**Comment by Coretool** [comment]
A ok...

**Comment by piranna** [comment]
> A ok...

You are more than welcome to work there, though :-) Just take one of the unchecked commands, implement it as a stream object and do a pull-request :-) It needs tests and docs too, so there's still a lot of work to do... :-)

**Comment by Coretool** [comment]
Well then ^^  I'll look into it

**Comment by piranna** [comment]
@lite20, any update on this? I tried to assign this issue to you but first you need to accept my invitation to join NodeOS devs team...

**Comment by lite20** [comment]
@piranna ah yes I believe I pushed my work to my fork. It's almost completed. If I didn't push it yet, I can go do so now

**Comment by lite20** [comment]
@piranna Oh, I haven't got the invite yet. Let me check again. (I'm very honoured, by the way)

**Comment by piranna** [comment]
And here you go :-)

**Comment by lite20** [comment]
I believe we finished this. If I'm wrong, how do I go about completing it now?

**Comment by piranna** [comment]
I think it got starved, it should not need too much time to finish it, mostly define an API about how the module for the creation of new users should work, since we have diferent approach here, you want a tarball with everything and I want a minimalist aproach, so we could investigate both.

Work on it and we'll discuss ;-)

**Comment by lite20** [comment]
Ah yes I remember now. You had some progress going. Is that on your personal fork?

**Comment by piranna** [comment]
I think so, don't remember.
